Objects/moduleobject.c

```diff
@@ -784,11 +786,88 @@ _PyModuleSpec_IsUninitializedSubmodule(PyObject *spec, PyObject *name)
     return rc;
 }
 
+static int
+_get_file_origin_from_spec(PyObject *spec, PyObject **origin)
+{
+    PyObject *has_location = NULL;
+    int rc = PyObject_GetOptionalAttr(spec, &_Py_ID(has_location), &has_location);
+    if (rc <= 0) {
+        return rc;
+    }
+    rc = PyObject_IsTrue(has_location);
+    Py_DECREF(has_location);
+    if (rc <= 0) {
+        return rc;
+    }
+    // has_location is true, so origin is a location
+    rc = PyObject_GetOptionalAttr(spec, &_Py_ID(origin), origin);
+    if (rc <= 0) {
+        return rc;
+    }
+    if (!PyUnicode_Check(*origin)) {
+        Py_DECREF(*origin);
+        *origin = NULL;
+        return 0;
+    }
```

Comment:
If `spec.origin` isn't there (a very unlikely case, I know) then I'd expect this code to crash.  Instead:

```suggestion
    if (*origin != NULL && !PyUnicode_Check(*origin)) {
        Py_DECREF(*origin);
        *origin = NULL;
        return 0;
    }
```

Reply:
I believe this is handled already, `PyObject_GetOptionalAttr` will return 0 in that case. I've added an assert + a test though!